src/internal/permissions.ts

```diff
@@ -0,0 +1,418 @@
+              function: "0x1::fungible_asset::revoke_permission",
+              functionArguments: [signer[0].borrow(), permission.asset],
+              typeArguments: [],
+            });
+          }
+          // TODO: object nft revoke
+          case PermissionType.NFT: {
+            return builder.add_batched_calls({
+              function: "0x1::object::revoke_permission",
+              functionArguments: [signer[0].borrow(), permission.assetAddress],
+              typeArguments: ["0x4::token::Token"],
+            });
+          }
+          default: {
+            console.log("Not implemented");
+            return Promise.resolve();
+          }
+        }
+      });
+
+      await Promise.all(permissionPromises);
+      return builder;
+    },
+  });
+
+  return transaction;
+}
+
+//  helper functions
+
+async function getPermissionedSigner(builder: AptosIntentBuilder, args: {
+  existingHandleAddress: string | null,
+  primaryAccount: SingleKeyAccount,
+  permissionedAccount: AbstractedEd25519Account
+}) {
+  if (args.existingHandleAddress) {
+    const signer = await builder.add_batched_calls({
+      function: "0x1::permissioned_delegation::permissioned_signer_by_key",
+      functionArguments: [BatchArgument.new_signer(0), args.permissionedAccount.publicKey.toUint8Array()],
+      typeArguments: [],
+    });
+    return { signer, isNewHandle: false };
+  }
+
+  // Create new handle and signer
+  const handle = await builder.add_batched_calls({
+    function: "0x1::permissioned_signer::create_storable_permissioned_handle",
+    functionArguments: [BatchArgument.new_signer(0), 360],
+    typeArguments: [],
+  });
+
+  const signer = await builder.add_batched_calls({
+    function: "0x1::permissioned_signer::signer_from_storable_permissioned",
+    functionArguments: [handle[0].borrow()],
+    typeArguments: [],
+  });
+
+  return { signer, handle, isNewHandle: true };
+}
+
+async function grantPermission(builder: AptosIntentBuilder, args: {
+  permissionedSigner: BatchArgument[],
+  permission: Permission
+}) {
+  switch (args.permission.type) {
+    case PermissionType.FungibleAsset:
+      return builder.add_batched_calls({
+        function: "0x1::fungible_asset::grant_permission",
+        functionArguments: [
+          BatchArgument.new_signer(0),
+          args.permissionedSigner[0].borrow(),
+          args.permission.asset, // do i need to convert this to AccountAddress? .... i guess not??
+          args.permission.balance,
+        ],
+        typeArguments: [],
+      });
+    case PermissionType.NFT: {
+      let txn: Promise<BatchArgument[]>[] = [];
+      const nftPermission = args.permission as NFTPermission; // bruh why do i need this
+      if (nftPermission.capabilities.transfer) {
+        // console.log("grant transfer");
+        // txn.push(builder.add_batched_calls({
+        //   function: "0x1::object::grant_permission",
+        //   functionArguments: [
+        //     BatchArgument.new_signer(0), 
```

Comment:
bruh

Reply:
I _think_ the types are getting messed up here due to `GenericPermission`. The `type` field of `GenericPermission` is `string` which also matches at the type level.